File: JavaScript/260-Single Number III.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function(nums) {
    let bitmask = 0;
    for (let num of nums){
        bitmask = bitmask^num;
    }
    
    const diff = bitmask&(-bitmask);
    let x = 0;
    for (let num of nums){
        if (num & diff){
            x = x^num;
        }
    }
    
    return [x, bitmask^x];
};

File: C++/260-Single Number III.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        long bitmask = 0;
        for (int num:nums) {
            bitmask ^= num;
        }

        long diff = bitmask&(-bitmask);  // use long here because the negation of -2147483648 cannot be represented in type 'int'
        int x = 0;
        for (int num:nums) {
            if (num&diff) {
                x = x^num;
            }
        }
        
        return {x, (int)bitmask^x};
    }
};